cluster/store/store_log.go

```diff
@@ -1,45 +1,83 @@
 
-import "github.com/hashicorp/raft"
+import (
+	"fmt"
+	"github.com/ByteStorage/FlyDB/config"
+	"github.com/ByteStorage/FlyDB/lib/datastore"
+	"github.com/hashicorp/raft"
+)
 
-// raftLog implements raft.LogStore interface, we can use it to store raft logs
-// how to store raft logs? we can use FlyDB/RocksDB/LevelDB/BoltDB to store raft logs
-// maybe we can use FlyDB
-type raftLog struct {
-	// implement me
-}
+// DataStoreFactory is a function type that creates a new instance of a raft.LogStore.
+// It takes a configuration map as input and returns the created LogStore or an error if the creation fails.
+type DataStoreFactory func(conf config.Options) (raft.LogStore, error)
 
-// newRaftLog returns a new raftLog
-func newRaftLog() raft.LogStore {
-	return &raftLog{}
-}
+// datastoreFactories is a map that associates a string key (name) with a DataStoreFactory.
+// It will be used to store and retrieve different DataStoreFactory implementations.
+var datastoreFactories = make(map[string]DataStoreFactory)
 
-func (r *raftLog) FirstIndex() (uint64, error) {
-	//TODO implement me
-	panic("implement me")
-}
+// Init initializes the datastoreFactories by registering different DataStoreFactory implementations.
+// It calls the Register function to associate each implementation with a unique name.
+// Currently, "memory" and "bolt" are registered as the names for the corresponding factories.
+// The function returns an error if any registration fails, but in this implementation, the error is ignored.
+func Init() error {
+	// Register the "memory" DataStoreFactory implementation with the name "memory"
+	_ = Register("memory", datastore.NewLogInMemStorage)
+	// Register the "bolt" DataStoreFactory implementation with the name "boltdb"
+	_ = Register("boltdb", datastore.NewLogBoltDbStorage)
+	// Register the "flydb" DataStoreFactory implementation with the name "flydb"
+	_ = Register("flydb", datastore.NewLogFlyDbStorage)
+	// Register the "rocksdb" DataStoreFactory implementation with the name "rocksdb"
+	_ = Register("rocksdb", datastore.NewLogRocksDbStorage)
 
-func (r *raftLog) LastIndex() (uint64, error) {
-	//TODO implement me
-	panic("implement me")
+	return nil
 }
 
-func (r *raftLog) GetLog(index uint64, log *raft.Log) error {
-	//TODO implement me
-	panic("implement me")
-}
+// newRaftLog is a function that returns a new instance of raft.LogStore.
+// It initializes the datastoreFactories by calling the Init function.
+// Then, it retrieves the "memory" DataStoreFactory from the datastoreFactories map using the "memory" string key.
+// Finally, it creates a new LogStore using the retrieved factory and an empty configuration map.
+// The created LogStore and an error (if any) are returned.
+func newRaftLog() (raft.LogStore, error) {
+	_ = Init()
 
-func (r *raftLog) StoreLog(log *raft.Log) error {
-	//TODO implement me
-	panic("implement me")
+	// Get the "memory" DataStoreFactory from the map
+	return getDataStore("memory", config.Options{})
```

Comment:
You can adds fields in the `config/cluster_Options.go` to provide user defined storage methods, then passes them in somewhere such as meta/store/region, and then passes them to the `newRaft` method, replacing the "memory" here.

Reply:
For example, you can store this field in the `store` in `func (s *store) newRaftNode()`, and then pass this field to the `newRaftLog` method. As for `config.Options{}`, you can remove it.

Reply:
@sjcsjc123 I changed the option and config; please check again.